File: problems/2444-fixed-bounds-subarrays/count_subarrays.cpp
/*
Solution for LeetCode problem 2444
Author: alex.does.leetcode@gmail.com
Last updated: 03 Apr 2024

See:
https://leetcode.com/problems/count-subarrays-with-fixed-bounds/
*/


#include <vector>
using std::vector;

class Solution {
  private:
    vector<int>* nums;
    int minK, maxK;
    bool black(int n) {
        return (*nums)[n] < minK || (*nums)[n] > maxK;
    }
    bool red(int n) {
        return (*nums)[n] == minK;
    }
    bool blue(int n) {
        return (*nums)[n] == maxK;
    }
  public:
    long long countSubarrays(vector<int>& nums, int minK, int maxK) {
        // Set up helper functions.
        this->nums = &nums;
        this->minK = minK;  this->maxK = maxK;
        // Initialize count
        long count = 0;
        int start = 0;
        while (true) {
            // Move start to the next white sequence
            while (start < nums.size() && black(start)) {
                start += 1;
            }
            if (start >= nums.size()) {
                // If there are no more segments, terminate
                break;
            }
            int end = start;
            // Special case when minK == maxK
            if (minK == maxK) {
              while (end < nums.size() && !black(end)) {
                end += 1;
              }
              long l = end - start;
              count += (l * (l+1)) / 2;
              start = end + 1;
              continue;
            }
            // Count how many fixed-bounds subarrays there are in this sequence
            int r = -1, b = -1;  // Initialize red and blue pointers
            while (true) {
                // Everytime we see a non-white block:
                if (end >= nums.size() || black(end) || red(end) || blue(end)) {
                    // 1. add to count
                    if (r != -1 && b != -1) {
                        int large = std::max(r, b);
                        int small = std::min(r, b);
                        count += (small-start+1) * (end-large);
                    }
                    if (end >= nums.size()) {
                      break;
                    }
                    // 2. Update red and blue pointers
                    if (red(end)) {
                        r = end;
                    } else if (blue(end)) {
                        b = end;
                    // 3 . This segment is complete, go on to next one
                    } else if (black(end)) {
                        break;
                    }
                }
                end += 1;
            }
            start = end + 1;
        }
        return count;
    }
};

int main(int, char*[]) {
    return 0;
}

File: problems/2444-fixed-bounds-subarrays/count_subarrays.py
'''
Solution for LeetCode problem 2444
Author: alex.does.leetcode@gmail.com
Last updated: 03 Apr 2024

See:
https://leetcode.com/problems/count-subarrays-with-fixed-bounds/
'''


class Solution:
    def countSubarrays(self, nums: List[int], minK: int, maxK: int) -> int:
      def black(n):
        return nums[n] < minK or nums[n] > maxK
      def red(n):
        return nums[n] == minK
      def blue(n):
        return nums[n] == maxK

      count = 0
      start = 0
      # Add a "bad" block at the end to simplify end conditions
      nums.append(minK-1)
      while True:
        # Move start to the next white sequence
        while start < len(nums) and black(start):
          start += 1
        if start == len(nums):
          # If there are no more white sequences, terminate
          break
        end = start
        # Special case when minK == maxK
        if minK == maxK:
          while not black(end):
            end += 1
          l = end - start
          count += (l * (l+1)) // 2
          start = end + 1
          continue
        # Count how many fixed-bounds subarrays there are in this sequence
        r = b = None  # Initialize red and blue pointers
        while True:
          # Everytime we see a non-white block:
          if black(end) or red(end) or blue(end):
            # 1. add to count
            if r != None and b != None:
              large = max(r, b)
              small = min(r, b)
              count += (small-start+1) * (end-large)
            # 2. Update red and blue pointers
            if red(end):
              r = end
            elif blue(end):
              b = end
            # 3 . This segment is complete, go on to next one
            elif black(end):
              break
          end +=1
        start = end + 1
      return count